astropy/table/table.py

```diff
@@ -3712,24 +3719,27 @@ def _rows_equal(self, other):
         if isinstance(other, Table):
             other = other.as_array()
 
-        if self.has_masked_columns:
-            if isinstance(other, np.ma.MaskedArray):
-                result = self.as_array() == other
-            else:
-                # If mask is True, then by definition the row doesn't match
-                # because the other array is not masked.
-                false_mask = np.zeros(1, dtype=[(n, bool) for n in self.dtype.names])
-                result = (self.as_array().data == other) & (self.mask == false_mask)
-        else:
-            if isinstance(other, np.ma.MaskedArray):
+        try:
+            self_is_masked = self.has_masked_columns
+            other_is_masked = isinstance(other, np.ma.MaskedArray)
+            if (self_is_masked + other_is_masked) == 1:
+                # remap variables to a and b where a is masked and b isn't
+                a, b = (
+                    (self.as_array(), other)
+                    if self_is_masked
+                    else (other, self.as_array())
+                )
+
                 # If mask is True, then by definition the row doesn't match
                 # because the other array is not masked.
-                false_mask = np.zeros(1, dtype=[(n, bool) for n in other.dtype.names])
-                result = (self.as_array() == other.data) & (other.mask == false_mask)
+                false_mask = np.zeros(1, dtype=[(n, bool) for n in a.dtype.names])
+                return (a.data == b) & (a.mask == false_mask)
             else:
-                result = self.as_array() == other
-
-        return result
+                return self.as_array() == other
+        except TypeError:
+            # numpy may complain that structured array are not comparable
+            # see https://github.com/astropy/astropy/issues/13421
+            return False
```

Comment:
There is also the potential for a `ValueError` if the tables are not broadcastable (this was also previously a warning).  E.g.
```
In [48]: np.array([1, 2]) == np.array([1, 2, 3])
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
Cell In[48], line 1
----> 1 np.array([1, 2]) == np.array([1, 2, 3])

ValueError: operands could not be broadcast together with shapes (2,) (3,) 
```
Even though it is a bit of duplication, you should try/except for `(TypeError, ValueError)` around *only* the two lines which actually do the comparison. Otherwise those broad catches could be masking real errors in all the other (non-trivial) code.

If there isn't a test for this case you should add it.

Reply:
Good thinking. I'm on it

Reply:
done